### PR TITLE
sd-ndisc: rewrite option parser

### PR DESCRIPTION
**Pull request by yuwata**
split-out of #30569.

### CODE REVIEW

File: src/libsystemd-network/ndisc-option.c [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+        assert(opt);
+        assert(options);
+
+        if (len < sizeof(struct nd_opt_rd_hdr) + sizeof(struct ip6_hdr))
+                return -EBADMSG;
+
+        if (opt[0] != SD_NDISC_OPTION_REDIRECTED_HEADER)
+                return -EBADMSG;
+
+        sd_ndisc_option *p = ndisc_option_new(SD_NDISC_OPTION_REDIRECTED_HEADER, offset);
```
Comment (evverx) [inline src/libsystemd-network/ndisc-option.c]: Looks like there is a memory leak here (and `p` isn't used by this function as far as I can see):
```sh
=================================================================
==259==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 560000 byte(s) in 10000 object(s) allocated from:
    #0 0x7fae7f0eb1bc in __interceptor_calloc (/usr/lib64/clang/16/lib/libclang_rt.asan-x86_64.so+0xeb1bc) (BuildId: 2a39bb4aa6d13ff866a4a47997cefdbe31a0350f)
    #1 0xad5b10 in ndisc_option_new /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:57:30
    #2 0xadbe98 in ndisc_option_parse_redirected_header /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:322:30
    #3 0xada07d in ndisc_parse_options /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-option.c:796:29
    #4 0xa76fab in ndisc_router_parse /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc-router.c:119:13
    #5 0xa6c9a8 in ndisc_handle_router /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:212:13
    #6 0xa6c0cb in ndisc_recv /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:261:24
    #7 0x7fae7e73b2bc in source_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4187:21
    #8 0x7fae7e7392e4 in sd_event_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4808:21
    #9 0x7fae7e73eddd in sd_event_run /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4869:21
    #10 0x7fae7e73fd69 in sd_event_loop /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4891:21
    #11 0x416f6d in run /home/vagrant/systemd-asan/build/../src/network/networkd.c:109:13
    #12 0x413ffd in main /home/vagrant/systemd-asan/build/../src/network/networkd.c:116:1
    #13 0x7fae7cc49b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #14 0x7fae7cc49c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #15 0x412a14 in _start (/usr/lib/systemd/systemd-networkd+0x412a14) (BuildId: 619afa9520903783d507f9c515898f3f7d7315e2)
SUMMARY: AddressSanitizer: 560000 byte(s) leaked in 10000 allocation(s).
systemd-networkd.service: Child 259 belongs to systemd-networkd.service.
```

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -1217,6 +1205,10 @@ static int ndisc_router_process_captive_portal(Link *link, sd_ndisc_router *rt)
                 ndisc_captive_portal_free(target);
         }
 
+        captive_portal = strdup(uri);
+        if (!captive_portal)
+                return log_oom();
```
Comment (evverx) [inline src/network/networkd-ndisc.c]: Looks like `networkd` crashes when it processes captive portals:
```sh
Program received signal SIGABRT, Aborted.
0x00007f69f16b0884 in __pthread_kill_implementation () from target:/lib64/libc.so.6
(gdb) bt
#0  0x00007f69f16b0884 in __pthread_kill_implementation () from target:/lib64/libc.so.6
#1  0x00007f69f165fafe in raise () from target:/lib64/libc.so.6
#2  0x00007f69f164887f in abort () from target:/lib64/libc.so.6
#3  0x00007f69f2b29b41 in log_assert_failed (text=0xb7de60 <str> "r > 0", file=0xb7eee3 "src/network/networkd-ndisc.c", line=1225, 
    func=0xb7e180 <__func__.ndisc_router_process_captive_portal> "ndisc_router_process_captive_portal") at ../src/basic/log.c:968
#4  0x000000000086f629 in ndisc_router_process_captive_portal (link=0x61c000000880, rt=0x607000059500) at ../src/network/networkd-ndisc.c:1225
#5  0x0000000000859ad3 in ndisc_router_process_options (link=0x61c000000880, rt=0x607000059500) at ../src/network/networkd-ndisc.c:1381
#6  0x000000000085158b in ndisc_router_handler (link=0x61c000000880, rt=0x607000059500) at ../src/network/networkd-ndisc.c:1662
#7  0x000000000084f450 in ndisc_handler (nd=0x60b000001640, event=SD_NDISC_EVENT_ROUTER, message=0x607000059500, userdata=0x61c000000880)
    at ../src/network/networkd-ndisc.c:1692
#8  0x0000000000a6e482 in ndisc_callback (ndisc=0x60b000001640, event=SD_NDISC_EVENT_ROUTER, message=0x607000059500)
    at ../src/libsystemd-network/sd-ndisc.c:42
#9  0x0000000000a6d80c in ndisc_handle_router (nd=0x60b000001640, packet=0x60b00000b3b0) at ../src/libsystemd-network/sd-ndisc.c:223
#10 0x0000000000a6c0cc in ndisc_recv (s=0x60f0000015d0, fd=23, revents=1, userdata=0x60b000001640) at ../src/libsystemd-network/sd-ndisc.c:261
#11 0x00007f69f313b2bd in source_dispatch (s=0x60f0000015d0) at ../src/libsystemd/sd-event/sd-event.c:4187
#12 0x00007f69f31392e5 in sd_event_dispatch (e=0x617000000080) at ../src/libsystemd/sd-event/sd-event.c:4808
#13 0x00007f69f313edde in sd_event_run (e=0x617000000080, timeout=18446744073709551615) at ../src/libsystemd/sd-event/sd-event.c:4869
#14 0x00007f69f313fd6a in sd_event_loop (e=0x617000000080) at ../src/libsystemd/sd-event/sd-event.c:4891
#15 0x0000000000416f6e in run (argc=1, argv=0x7ffecf5596c8) at ../src/network/networkd.c:109
#16 0x0000000000413ffe in main (argc=1, argv=0x7ffecf5596c8) at ../src/network/networkd.c:116
```

**Comment by yuwata** [comment]
@evverx Thanks. Fixed the issues now.

**Comment by evverx** [comment]
Looks like it no longer crashes and the memory leak is gone too.

I read the commit messages and I have to admit I'm not sure why the parser is rewritten. Does it have anything to do with neighbor advertisements in the sense that it makes it easier to reuse the parser there too? Or is it something else entirely?

**Comment by yuwata** [comment]
> Looks like it no longer crashes and the memory leak is gone too.

Yay!

> I read the commit messages and I have to admit I'm not sure why the parser is rewritten. Does it have anything to do with neighbor advertisements in the sense that it makes it easier to reuse the parser there too?

Right. The parser will be also used parsing other types of NDisc message, e.g. Neighbor Advertisement.

**Review by bluca — APPROVED** [review]

**Pull request merged**